### PR TITLE
DB-10360 support operation on current_timestamp

### PR DESCRIPTION
**Pull request by yxia92**

### CODE REVIEW

**Comment by yxia92** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1774/

**Comment by yxia92** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1775/

**Comment by yxia92** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by jyuanca — APPROVED** [review]

**Review by ipraznik-splice — APPROVED** [review]

**Review by arnaud-splice — APPROVED** [review]

**Review by hatyo — APPROVED** [review]
It is amazing that you managed to provide this functionality without having to implement the `INTERVAL` data type, I think we might be find it also useful for defining the minimum retention period.
Great work!!

**Review by ascend1 — COMMENTED** [review]
1. The requirement definition is blurry because it says only "DB2-like operations on current_timestamp". Not surprisingly, DB2 supports interval values here:
```
values current timestamp + interval('1 day 3 hours');
1                         
------------------------- 
2020-10-28 12:27:14.42119 
```
Since we don't have interval syntax, it's not easy for us to have a complete "DB2-like" support for this. Maybe we should check if the fix is sufficient for the customer. I left this comment to the Jira ticket, too.

2. Now we have HOURS, MINUTES, and SECONDS, the following statement becomes valid:
```
splice> select timestamp('2020-01-01 00:00:00') + 2 minutes;
ERROR 2202F: Invalid time span operation.
```
At the moment it throws invalid time span operation, which could be confusing.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj [hunk 1]
```diff
@@ -7040,7 +7078,8 @@ convertAddTimeIntervalToFunction(ValueNode base, TimeSpanNode timeSpanNode, bool
                 function = "ADD_YEARS";
```
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj:7078]: To echo Zhen's comment: We should probably support hours/minutes/seconds in this code path as well.
Alternatively, why not merge the two code paths? They seem redundant.

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by yxia92** [comment]
Thanks @hatyo @ascend1 and @arnaud-splice for the comment. In this JIRA, I only extended the support for current_timestamp, but not any operation on  general timestamp value/columns. So for the example that @ascend1 listed:
```
splice> select timestamp('2020-01-01 00:00:00') + 2 minutes;
ERROR 2202F: Invalid time span operation.
```
it is not supported.
This is because at the time we parse the expression, syntaxer does not know about the type of the expression. So our syntax rules only expect 3 categories:
1. regular +/- operation for any data type without the time span units; or
2. Current_timestamp node +/- value node followed by time span units, this will be converted to timestampadd() functions; or otherwise,
3. value/column +/- value node followed by time span units except for #2, the value/column will be assumed as a DateType. and for date type, the time span unit can only support day/month/year, but not minute/hour/second.
If needed, we can have a separate JIRA to define the scope of general support.

**Pull request merged**